stdlib/public/runtime/SymbolInfo.cpp

```diff
@@ -17,25 +17,88 @@
 #define NOMINMAX
 #include <Windows.h>
 #include <DbgHelp.h>
+#include <Psapi.h>
+#include <shlwapi.h>
 #elif SWIFT_STDLIB_HAS_DLADDR
 #include <dlfcn.h>
 #endif
 
+#include <array>
+#include <cstdlib>
+
 #include "ImageInspection.h"
+#include "swift/Runtime/Win32Strings.h"
 
 using namespace swift;
 
-const char *SymbolInfo::getFilename() const {
+SymbolInfo::SymbolInfo()
 #if defined(_WIN32) && !defined(__CYGWIN__)
-  return nullptr;
+: _package({}),
+#elif SWIFT_STDLIB_HAS_DLADDR
+: _info({}),
+#endif
+_address(nullptr)
+{
+}
+
+SymbolInfo::~SymbolInfo() {
+#if defined(_WIN32) && !defined(__CYGWIN__)
+  if (char *imagePath = _imagePath.getValueOr(nullptr)) {
+    free(imagePath);
+  }
+#endif
+}
+
+const void *SymbolInfo::getAddress() const {
+  return _address;
+}
+
+const char *SymbolInfo::getImagePath() const {
+#if defined(_WIN32) && !defined(__CYGWIN__)
+  if (!_imagePath.has_value()) {
+    // On Win32/Win64, the base address of a loaded image and its HMODULE are
+    // the same pointer, so we can simply cast from one to the other.
+    HMODULE hModule = reinterpret_cast<HMODULE>(getImageBaseAddress());
+    if (!hModule) {
+      _imagePath = nullptr;
+      return;
+    }
+
+    // Call GetModuleFileNameW() instead of SymGetModuleInfoW() because it
+    // doesn't require the use of the global DbgHelp lock and doesn't fetch
+    // other information we don't need.
+    std::array<wchar_t, MAX_PATH> filename;
+    if (GetModuleFileNameW(hModule, filename.data(), filename.size())) {
```

Comment:
Just FWIW, the Win32 docs suggest that `GetModuleFileNameW()` will fail if the module list isn't initialised yet, among other things. That's why I went with `GetMappedFileNameW()` instead.